spring-pulsar-test/src/main/java/org/springframework/pulsar/test/support/PulsarTestConsumer.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.pulsar.test.support;
+
+import static org.awaitility.Awaitility.await;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.pulsar.client.api.Consumer;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.client.api.SubscriptionInitialPosition;
+import org.awaitility.core.ConditionTimeoutException;
+
+import org.springframework.pulsar.core.PulsarConsumerFactory;
+
+/**
+ * Fluent API, to be used in tests, for consuming messages from Pulsar topics until a
+ * certain {@code Condition} has been met.
+ *
+ * @param <T> the type of the message payload
+ * @author Jonas Geiregat
+ */
+public class PulsarTestConsumer<T> implements TopicStep, SchemaStep, ConditionsStep<T> {
+
+	private final PulsarConsumerFactory<T> consumerFactory;
+
+	private List<String> topics;
+
+	private Condition<T> condition;
+
+	private Schema<T> schema;
+
+	private Duration timeout = Duration.ofSeconds(30);
+
+	public static <T> TopicStep consumeMessages(PulsarConsumerFactory<T> pulsarConsumerFactory) {
+		return new PulsarTestConsumer<>(pulsarConsumerFactory);
+	}
+
+	public PulsarTestConsumer(PulsarConsumerFactory<T> consumerFactory) {
+		this.consumerFactory = consumerFactory;
+	}
+
+	@Override
+	public PulsarTestConsumer<T> fromTopic(String... topics) {
+		this.topics = List.of(topics);
+		return this;
+	}
+
+	@Override
+	public ConditionsStep<T> awaitAtMost(Duration timeout) {
+		this.timeout = timeout;
+		return this;
+	}
+
+	@Override
+	public List<Message<T>> until(Condition<T> condition) {
+		this.condition = condition;
+		try {
+			return doGet();
+		}
+		catch (ConditionTimeoutException ex) {
+			throw new PulsarTimeOutException("Timeout waiting for messages", ex);
+		}
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <E> ConditionsStep<E> withSchema(Schema<E> schema) {
+		this.schema = (Schema<T>) schema;
+		return (ConditionsStep<E>) this;
+	}
+
+	private List<Message<T>> doGet() {
```

Comment:
So after digging in a bit, I think my initial idea of using Awaitility ended up being more trouble than its worth. Mostly because in order to support the "grab as many messages as you can in N time units" use case is super clunky to do. 

We should be able to do the following: 
```java
PulsarConsumerFactory<UserPojo> cf = null;

// Get 10 w/in 5 seconds or go boom
List<Message<UserPojo>> users = PulsarTestConsumer.consumeMessages(cf)
		.fromTopic("user-topic")
		.withSchema(Schema.JSON(UserPojo.class))
		.awaitAtMost(Duration.ofSeconds(5))
		.until(Conditions.desiredMessageCount(10))
		.get();

// Get as many as you can in 5 seconds
users = PulsarTestConsumer.consumeMessages(cf)
		.fromTopic("user-topic")
		.withSchema(Schema.JSON(UserPojo.class))
		.awaitAtMost(Duration.ofSeconds(5))
		.get();
```

We will need to loosen the ConditionsStep so that until() also returns a step and then we add a method to terminate and get the messages. This way we are not forced to provide a condition if we simply want all messages we can get in N time units.

I believe we can replace everything from line 91 down w/ the following (similar to your original approach):

```java
	@Override
	public List<Message<T>> get() {
		var messages = new ArrayList<Message<T>>();
		try (Consumer<T> consumer = consumerFactory.createConsumer(this.schema, this.topics, "test-consumer",
				c -> c.subscriptionInitialPosition(SubscriptionInitialPosition.Earliest))) {
			while (timeRemaining()) {
				var message = consumer.receive();
				messages.add(message);
				consumer.acknowledge(message);
				if (this.condition != null) {
					if (this.condition.meets(messages)) {
						return messages;
					}
				}
				Thread.sleep(200);
			}
		}
		catch (PulsarClientException | InterruptedException ex) {
			throw new PulsarException(ex);
		}
		if (this.condition != null && !this.condition.meets(messages)) {
			throw new ConditionTimeoutException("things go here");
		}
		return messages;
	}

	private boolean timeRemaining() {
	   // todo ;
	}
```

Wdyt?

Reply:
I can follow your line of reasoning and will remove awaitility